### PR TITLE
os/bluestore: prevent lock for almost "flush" calls

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -6471,6 +6474,7 @@ void BlueStore::_txc_finish(TransContext *txc)
   for (set<OnodeRef>::iterator p = txc->onodes.begin();
        p != txc->onodes.end();
        ++p) {
+    (*p)->flushing_count--;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Isn't it better to decrement the counter AFTER flush_txcs processing?

**Comment by ifed01** [comment]
+1

**Comment by liewegas** [comment]
Conflicts with https://github.com/ceph/ceph/pull/12541

**Comment by varadakari** [comment]
@yuyuyu101 could you please resolve the conflicts?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -6411,6 +6413,7 @@ void BlueStore::_txc_write_nodes(TransContext *txc, KeyValueDB::Transaction t)
 
     std::lock_guard<std::mutex> l(o->flush_lock);
     o->flush_txns.insert(txc);
+    o->flushing_count++;
   }
 
   // objects we modified but didn't affect the onode
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: need to increment flushing_count below, too

**Comment by liewegas** [comment]
<pre>
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BlueStore.cc: In member function ‘void BlueStore::_txc_write_nodes(BlueStore::TransContext*, KeyValueDB::Transaction)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BlueStore.cc:6453:7: error: ‘o’ was not declared in this scope
       o->flushing_count++;
       ^
</pre>

**Comment by yuyuyu101** [comment]
done

**Comment by liewegas** [comment]
needs rebase, sorry!

**Pull request merged**